File: vol2/ch21/LeetCode474.cpp
class Solution {
public:
    pair<int,int> Count(string&str) {
        int len=str.length();
        int cnt0=0,cnt1=0;
        for (int i=0;i<len;i++) {
            if(str[i]=='1') cnt1++;
            else cnt0++;
        }
        return pair<int,int>(cnt0,cnt1);
    }
    int findMaxForm(vector<string>& strs,int m,int n) {
        int len=strs.size();
        vector<vector<int>> dp(m+1,vector<int>(n+1,0));
        for (int i=1;i<=len;i++) {
            auto [cnt0,cnt1]=Count(strs[i-1]);
            for (int j=m;j>=cnt0;j--) {
                for (int k=n;k>=cnt1;k--)
                    dp[j][k]=max(dp[j][k],dp[j-cnt0][k-cnt1]+1);
            }
        }
        return dp[m][n];
    }
};

File: vol2/ch21/LeetCode474.py
class Solution:
    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        length=len(strs)
        dp=[[0 for _ in range(n+1)] for _ in range(m+1)]
        for i in range(1,length+1):
            [cnt0,cnt1]=self.Count(strs[i-1])
            for j in range(m,cnt0-1,-1):
                for k in range(n,cnt1-1,-1):
                    dp[j][k]=max(dp[j][k],dp[j-cnt0][k-cnt1]+1)
        return dp[m][n]
    def Count(self,str):
        length=len(str)
        cnt0,cnt1=0,0
        for i in range(0,length):
            if str[i]=='1':cnt1+=1
            else: cnt0+=1
        return [cnt0,cnt1]
